akka-docs/rst/java/code/docs/stream/javadsl/cookbook/RecipeGlobalRateLimit.java

```diff
@@ -121,11 +119,12 @@ private Limiter(int maxAvailableTokens, FiniteDuration tokenRefreshPeriod,
 
     private void releaseWaiting() {
       final List<ActorRef> toBeReleased = new ArrayList<>(permitTokens);
-      for (int i = 0; i < permitTokens && i < waitQueue.size(); i++) {
-        toBeReleased.add(waitQueue.remove(i));
+      for (Iterator<ActorRef> it = waitQueue.iterator(); permitTokens > 0 && it.hasNext();) {
+          toBeReleased.add(it.next());
+          it.remove();
+          permitTokens --;
       }
 
-      permitTokens -= toBeReleased.size();
```

Comment:
If the `waitQueue` contains 2 elements and `permitTokens=2`, only one element is removed. Because `waitQueue.remove(i)` decreases the size of the `waitQueue` and in the second iteration the result of the `i < waitQueue.size()` expression becomes false.